README.md
# PyRadio

Command line internet radio player.

Ben Dowling - [https://github.com/coderholic](https://github.com/coderholic)

## Table of contents

* [Requirements](#requirements)
* [Installation](#installation)
* [Command line options](#command-line-options)
* [Controls](#controls)
* [PyRadio's Modes](#pyradios-modes)
* [Config file](#config-file)
* [About Playlist files](#about-playlist-files)
* [Search function](#search-function)
* [Line editor](#line-editor)
* [Moving stations around](#moving-stations-around)
* [Specifying stations' encoding](#specifying-stations-encoding)
* [Player detection / selection](#player-detection-selection)
* [Player default volume level](#player-default-volume-level)
* [Displaying Station Info](#displaying-station-info)
* [Copying and pasting - Registers](#copying-and-pasting---registers)
* [PyRadio Themes](#pyradio-themes)
* [Session Locking](#session-locking)
* [Update notification](#update-notification)
* [Cleaning up](#cleaning-up)
* [Debug mode](#debug-mode)
* [Reporting bugs](#reporting-bugs)
* [Acknowlegement](#acknowlegement)

## Requirements

* python 2.7+/3.5+
* MPV, MPlayer or VLC installed and in your path.

## Installation

The best way to install **PyRadio** is via a distribution package, if one exists (e.g. *Arch Linux* and derivatives can install [pyradio-git](https://aur.archlinux.org/packages/pyradio-git/) from AUR).

In any other case, and since **PyRadio** is currently not available via pip, you will have to [build it from source](build.md).

## Command line options

```
$ pyradio -h

usage: pyradio [-h] [-s STATIONS] [-p [PLAY]] [-u USE_PLAYER] [-a] [-ls] [-l]
               [-t THEME] [-scd] [-ocd] [-d]

Curses based Internet radio player

optional arguments:
  -h, --help            show this help message and exit
  -s STATIONS, --stations STATIONS
                        Use specified station CSV file.
  -p [PLAY], --play [PLAY]
                        Start and play.The value is num station or empty for
                        random.
  -u USE_PLAYER, --use-player USE_PLAYER
                        Use specified player. A comma-separated list can be
                        used to specify detection order. Supported players:
                        mpv, mplayer, vlc.
  -a, --add             Add station to list.
  -ls, --list-playlists
                        List of available playlists in config dir.
  -l, --list            List of available stations in a playlist.
  -t THEME, --theme THEME
                        Use specified theme.
  -scd, --show-config-dir
                        Print config directory location and exit.
  -ocd, --open-config-dir
                        Open config directory with default file manager.
  --unlock              Remove sessions' lock file.
  -d, --debug           Start pyradio in debug mode.
```

The following options can also be set in **PyRadio**'s [configuration file](#config-file):

* **-s** - parameter **default_playlist** (default value: **stations**)
* **-p** - parameter **default_station** (default value: **-1**)
* **-u** - parameter **player** (default value: **mpv, mplayer, vlc**)
* **-t** - parameter **theme** (default value: **dark**)

## Controls

```
                  Main window                                      Playlists window                   Themes window
-------------------------------------------------------------------------------------------------------------------------------------
Up/Down/j/k/
PgUp/PgDown       Change station selection                         Change station playlist            Change station theme
g                 Jump to first station                            Jump to first playlist             Jump to first theme
<n>G              Jump to n-th / last station                      Jump to n-th / last playlist       Jump to n-th / last theme
H M L             Jump to the top / middle bottom of the list      [Valid]                            -
P                 Jump to playing station                          Jump to playing playlist           -
Enter/Right/l     Play selected station                            Open selected playlist             Apply selected theme
r                 Select and play a random station                 Re-read playlists from disk        -
Space/Left/h      Stop/start playing selected station              -                                  -
Space             -                                                -                                  Apply theme and make it default
-/+ or ,/.        Change volume                                    [Valid]                            [Valid]
m                 Mute / unmute player                             [Valid]                            [Valid]
v                 Save volume (not applicable for vlc)             [Valid]                            [Valid]
o s R             Open / Save / Reload playlist                    -                                  -
a A               Add / append a new station                       -                                  -
e                 Edit current station                             -                                  -
E                 Change station's encoding                        -                                  -
DEL,x             Delete selected station                          -                                  -
t T               Load theme / Toggle transparency                 [Valid]                            [Valid]
c                 Open Configuration window.                       -                                  -
/ n N             Search, go to next / previous result             [Valid]                            [Valid]
J                 Create a jump tag
<n>^U <n>^D       Move station up / down.                          -                                  -
?                 Show keys help                                   [Valid]                            [Valid]
#                 Redraw window                                    [Valid]                            [Valid]
Esc/q             Quit                                             -                                  -
Esc/q/Left/h      -                                                Cancel / close window              Cancel / close window
```

The same logic applies to all **PyRadio** windows.

**Note:** All windows - except the *Search window* - support changing the volume and muting / unmuting the player (provided that **PyRadio** is actually connected to a station).


## PyRadio's Modes

**PyRadio** has the following primary modes:

1. The **Main** mode, which is the one you get when you open the program, showing you a list of stations (a playlist), that you can play and edit; this is why it is also called the **editing mode**. All other modes derive from this one, and it's the mode you have to get to in order to terminate the program.

2. The **Playlist** mode, which you can open by pressing "***o***". Then you can open, create, paste a station, etc.

3. The **Registers** mode. This is identical to the "*Playlist*" mode, but instead of displaying playlists, it displays register. You can enter this mode by pressing "***''***" (two single quotes) and exit from it by pressing "***Esc***" or "***q***".

4. The **Register Main** mode, which is identical to the "*Main*" mode, except it displays the content of a **named** register.

A set of **secondary modes** is also available (a secondary mode works within a primary one):

1. The **Extra Commands** mode, which gives you access to extra commands. You can enter this mode by pressing "***\\***" (backslash). Then a backslash is displayed at the bottom right corner of the window.

2. The **Yank (Copy)** mode, which is used to copy stations to **registers**. You can enter this mode by pressing "***y***". Then a "*y*" is displayed at the bottom right corner of the window.

3. The **Open Register** mode, which is used to open a register or get into the *Registers* mode. You can enter this mode by pressing "***'***" (single quote). Then a single quote is displayed at the bottom right corner of the window.

The functions availabe through the *secondary modes* are content dependant, so you can see what command is available by pressing "***?***" while within a secondary mode. Pressing any other key will exit the secondary mode.

## Config file

**PyRadio** upon its execution tries to read its configuration file (i.e. *~/.config/pyradio/config*). If this file is not found, it will be created. If an error occurs while parsing it, an error message will be displayed and ***PyRadio*** will terminate.

The file contains parameters such as the player to use, the playlist to load etc. It is heavily commented (as you can see [here](pyradio/config)), so that manual editing is really easy. The best practice to manually edit this file is executing ***PyRadio*** with the **-ocd** command line option, which will open the configuration directory in your file manager, and then edit it using your preferable text editor.

The file can also be altered while **PyRadio** is running by pressing "***c***", which will open the "***Configuration window***". This window presents all **PyRadio** options and provide the way to change them and finally save them by pressing "***s***".

In any case, **PyRadio** will save the file before exiting (or in case Ctrl-C is pressed) if needed (e.g. if a config parameter has been changed during its execution).

If saving the configuration file fails, **PyRadio** will create a back up file and terminate. When restarted, **PyRadio** will try to restore previously used settings from the said back up file.

## About Playlist files

**PyRadio** reads the stations to use from a CSV file, where each line contains two columns, the first being the station name and the second being the stream URL.

Optionally, a third column can be inserted, stating the encoding used by the station (more on this at [Specifying stations' encoding](#specifying-stations-encoding)).

**PyRadio** will by default load the user's stations file (e.g. *~/.config/pyradio/stations.csv*) to read the stations from. If this file is not found, it will be created and populated with a default set of stations.

**Tip:** If you already have a custom *stations.csv* file, but want to update it with **PyRadio**'s default one, you just rename it, run **PyRadio** (so that the default one get created) and then merge the two files.


**Note:** Older versions used to use ***~/.pyradio*** as default stations file. If this file is found, it will be copied to use's config directory (e.g. ***~/.config/pyradio***) and renamed to ***stations.csv*** or if this file exists, to ***pyradio.csv***. In this case, this file will be the default one.

### Specifying a playlist to load (command line)

**PyRadio** will normally load its default playlist file, as described above, upon its execution. A different file can be loaded when the **-s** command line option is used.

The ***-s*** option will accept:

* a relative or absolute file name.
* the name of a playlist file which is already in its configuration directory.
* the number of a playlist file, as provided by the ***-ls*** command line option.

Examples:

To load a playlist called "***blues.csv***", one would use the command:

```
pyradio -s /path/to/blues.csv
```

If this file was saved inside ***PyRadio***'s configuration directory, one could use the following command:

```
pyradio -s blues
```

To use the playlist number, one would execute the commands:

```
$ pyradio -ls
Playlists found in "/home/user/.config/pyradio"
  1. hip-hop
  2. party
  3. stations
  4. huge
  5. blues
  6. rock
  7. pop
$ pyradio -s 5
```

**Note:** The default playlist to load can also be set in **PyRadio**'s [configuration file](#config-file), parameter **default_playlist** (default value is ***stations***).

### Managing playlists (within PyRadio)

Once **PyRadio** has been loaded, one can perform a series of actions on the current playlist and set of playlists saved in its configuration directory.

Currently, the following actions are available:

Pressing "***a***" or "***A***" will enable you to add a new station (either below the currently selected station or at the end of the list), while "***e***" will edit the currently selected station. All of these actions will open the "*Station Editor*".

If you just want to change the encoding of the selected station, just press "***E***". If the station is currently playing, playback will be restarted so that the encoding's change takes effect (hopefully correctly displaying the station/song title).

Then, when this is done, you can either save the modified playlist, by pressing "***s***", or reload the playlist from disk, by pressing "***R***". A modified playlist will automatically be saved when **PyRadio** exits (or Ctrl-C is pressed).

One thing you may also want to do is remove a station from a playlist, e.g. when found that it not longer works. You can do that by pressing "***DEL***" or "***x***". The deleted station is copied to the **unnamed register** (refer to section [Copying and pasting - Registers](#copying-and-pasting---registers) for more information).

Finally, opening another playlist is also possible. Just press "***o***" and you will be presented with a list of saved playlists to choose from. These playlists must be saved beforehand in **PyRadio**'s configuration directory.

While executing any of the previous actions, you may get confirmation messages (when opening a playlist while the current one is modified but not saved, for example) or error messages (when an action fails). Just follow the on screen information, keeping in mind that a capital letter as an answer will save this answer in **PyRadio**'s configuration file for future reference.

### Managing "foreign" playlists

A playlist that does not reside within the program's configuration directory is considered a "***foreign***" playlist. This playlist can only be opened by the "***-s***" command line option.

When this happens, **PyRadio** will offer you the choice to copy the playlist in its configuration directory, thus making it available for manipulation within the program.

If a playlist of the same name already exists in the configuration directory, the "***foreign***" playlist will be time-stamped. For example, if a "***foreign***" playlist is named "*stations.csv*", it will be named "*2019-01-11_13-35-47_stations.csv*" (provided that the action was taken on January 11, 2019 at 13:35:47).

### Playlist history

**PyRadio** will keep a history of all the playlists opened (within a given session), so that navigating between them is made easy.

In order to go back to the previous playlist, the user just has to press "***\\\\***" (double backslash). To get to the first playlist "***\\]***" (backslash - closing square bracket) can be used.

Going forward in history is not supported.

## Search function

On any window presenting a list of items (stations, playlists, themes) a **search function** is available by pressing "***/***".

The *Search Window* supports normal and extend editing and in session history.

One can always get help by pressing the "***?***" key.

After a search term has been successfully found (search is case insensitive), next occurrence can be obtained using the "***n***" key and previous occurrence can be obtained using the "***N***" key.

## Line editor

**PyRadio** "*Search function*" and "*Station edior*" use a *line editor* to permit typing and editing stations' data.

The *line editor* works both on **Python 2** and **Python 3**, but does not provide the same functionality for both versions:


* In **Python 2**, only ASCII characters can be inserted.
* In **Python 3**, no such restriction exists. Furthermore, using CJK characters is also supported.

One can always display help by pressing "***?***", but that pauses a drawback; one cannot actually have a "**?**" withing the string.

To do that, one would have to use the backslash key "**\\**" and then press "**?**".

To sum it all up:

1. Press "***?***" to get help.
2. Press "***\\?***" to get a "**?**".
3. Press "***\\\\***" to get a "**\\**".

### CJK characters support

The *line editor* supports the insertion of [CJK Unified Ideographs](https://en.wikipedia.org/wiki/CJK_Unified_Ideographs), as described on [CJK Unified Ideographs (Unicode block)](https://en.wikipedia.org/wiki/CJK_Unified_Ideographs_(Unicode_block)) also known as URO, abbreviation of Unified Repertoire and Ordering. These characters, although encoded as a single code-point (character), actually take up a 2-character space, when rendered on the terminal.

A depiction of the editor's behavior can be seen at this image: [pyradio-editor.jpg](https://members.hellug.gr/sng/pyradio/pyradio-editor.jpg).

## Moving stations around

Rearranging the order of the stations in the playlist is another feature **PyRadio** offers.

All you have to do is specify the *source* station (the station to be moved) and the position it will be moved to (*target*).

There are three way to do that:

1. Press **Ctrl-U** or **Ctrl-D** to move the current station up or down.
2. Type a station number and press **Ctrl-U** or **Ctrl-D** to move the current station there.
3. Go to the position you want to move a station to, and press "**J**". This will *tag* this position (making it the *target* of the move). Then go to the station you want to move and press **Ctrl-U** or **Ctrl-D** to move it there.

## Specifying stations' encoding

Normally, stations provide information about their status (including the title of the song playing, which **PyRadio** displays) in Unicode (**utf-8** encoded). Therefore, **PyRadio** will use **utf-8** to decode such data, by default.

In an ideal world that would be the case for all stations and everything would be ok and as far as **PyRadio** is concerned, songs' titles would be correctly displayed. Unfortunately, this is not the case.

A lot of stations encode and transmit data in a different encoding (typically the encoding used at the region the come from). The result in **PyRadio** would be that a song title would be incorrectly displayed, not displayed at all, or trying to displaying it might even break **PyRadio**'s layout.

**Note:** **vlc** will not work in this case; it presumably tries to decode the said data beforehand, probably using **utf-8** by default, and when it fails, it provides a "**(null)**" string, instead of the actual data. So, you'd better not use **vlc** if such stations are in your playlists.

**PyRadio** addresses this issue by allowing the user to declare the encoding to use either in a station by station mode or globally.

### Station by station encoding declaration

As previously stated, a **PyRadio**'s playlist can optionally contain a third column (in addition to the station name and station URL columns), which declares the station's encoding.

So, when a **non-utf-8** encoded station is inserted in a playlist, its encoding can also be declared along with its other data. The drawback of this feature is that an encoding must be declared for **all stations** (so that the ***CSV*** file structure remains valid). To put it simple, since one station comprises the third column, all stations must do so as well.


Changelog
2020-01-26 s-n-g
    * 0.8.8-beta1
    * Adding connection timeout counter
    * Adding station info window
    * Adding registers add, remove station, clear
    * Adding paste station to current playlist
    * g prefixed by a number jumps to it (same as G)
    * Dead keys (alternative mode and jump number) are
      now displayed at the bottom right corner of the window
    * Main help window has 3 pages nows
    * Fixing playback restart when encoding changed
    * Replacing widechar.py with cjkwrap.py and adding its
      update script in devel/

2019-12-23 s-n-g
    * Version 0.8.7
    * Fixing volume issue with mpv
    * mpv on python3 uses socket communication only; no stdout parsing
      done anymore, as it is still done on python2, due to title
      (icy-title) encoding problems.
    * socat is no longer needed to use mpv

2019-12-14 s-n-g
    * Version 0.8.6
    * Adding playlist history (for local playlists)
    * https URLs will be converted to http before connecting
    * Fixing station moving when appending station
    * Config / Default station: pading fixed
    * WINDOWS: Volume will be saved when mplayer is installed in %APPDATA%\pyradio

2019-11-15 s-n-g
    * PyRadio will not crush with mpv 0.30.0
      Changing mpv's volume is still possible, but no info will be
      presented on the Status Bar.
      Furthermore, saving mpv's volume will not be possible
      ( mpv issue #7153: https://github.com/mpv-player/mpv/issues/7153 )

2019-11-12 s-n-g
    * When default played is changed in the config, a message to restart
      the application is presented to the user
    * Config / Default station: pading fixed

2019-11-10 s-n-g
    * Fixing vlc returned volume parsing (due to locales decimal separator)

2019-10-29 s-n-g
    * FIX: Playlists flagged as changed when adding a station
    * Handling CJK presentation on station and playlist view
    * Preparing for online stations browser

2019-10-23 s-n-g
    * Handling CJK presentation on station and playlist window

2019-10-20 s-n-g
    * Version 0.8.2
    * Fixing chars H,L,M rejected by line editor
    * Fixing station editor rejecting "\?"
    * Updating station's editor help messages

2019-10-19 s-n-g
    * Version 0.8.1
    * CJK Unified Ideographs supported by the line editor
    * On python 2, trying to edit a station whose name contains
      non-ASCII characters is prohibited and will end up in
      displaying a relevant message
    * Line editor: using backslash to insert "?" and "\"
    * Search term will not be lost when resizing the window
    * Fixing issues with presenting search history
    * Themes now have a "Edit Cursor" field

2019-09-08 s-n-g
    * Version 0.8.0
    * Adding station editor ("a" and "A" to add a station, "e" to edit)
    * Line editor supports unlimited string length
    * Main help window separated to two pages (navigation with "n" / "p")
    * Changing "e" to "E" to change a station's encoding
    * Changing "p" to "P" to jump to playing station / loaded playlist
    * Adding H, L to jump to top / bottom of screen
    * Changing M to jump to middle of screen
    * Changing volume, saving volume and muting is now available on most
      windows (pop up and questions)
    * Manipulating volume (keys m,v) on a help window, will close it
      if player not playing
    * Adding ^U, ^D to move station up, down
    * Search string will not be lost after displaying help
    * PyRadio runs on Windows (finally). Added an installation BAT file,
      icons, program shortcut and help (windows.md - windows.html)
    * Minor bug fixes

2019-07-15 s-n-g
    * Searching is now available on any window presenting a list of items
    * Transparency indication presentation always reflects setting
    * A locked session is indicated next to PyRadio version
    * Adding "Top" link to html files
    * Docs updated

2019-07-08 s-n-g

    * Version 0.7.8
    * fixing playlist recovery message presentation
    * fixed foreign playlist management
    * fixed exiting when playlist is modified
    * when applying a them which is not supported, using "light"
      if "light" or "light_16_colors" is default
    * fixing update notification message "Will check again in 0 days"
    * update notification will clean files in all cases
    * adding "M" to config station selection
    * updating help for playlist and config station selection
    * heavy refactoring; using window stack and redisplay list

2019-06-26 s-n-g
    * Version 0.7.7
    * mpv now uses a dedicated socket file. This way multiple instances of
      PyRadio can be executed.
    * Introducing session locking.
    * Added the "--unlock" command line parameter, to force sessions' unlock.
    * Added "M" command, which will jump to the middle of the list.
    * PyRadio can load external theme files.
    * Three more themes added. These are system themes (actual files).
    * Theme selection window reworked - themes are separated by location,
      theme selection is remembered when resizing, and loading default or
      saved theme (in config window).
    * PyRadio will report reverting to default theme.
    * PyRadio will check and report when a new release is available.
    * Added good bye message.
    * Theme editor implementation started (disabled for this release).
    * Minor other fixes.

2019-06-06 s-n-g
    * Verion 0.7.6.2
      This is a BUG FIX release, fixing config status (indicating whether
      config is modified or not)

2019-06-02 s-n-g
    * Version 0.7.6.1
      This is a BUG FIX release, fixing regression: config not saved when
      changing themes

2019-06-01  s-n-g
    * Version 0.7.6
    * Added "e" option to change station's encoding.
    * Implemented playlist backup and recovery, to address saving errors.
    * Inform user when playlist not found and of playlist recovery result.
    * Parameter -s - Check if file is supported (ends with .csv)

2019-04-18  s-n-g
    * Version 0.7.5

pyradio/main.py
# -*- coding: utf-8 -*-
import sys
import curses
import logging
from argparse import ArgumentParser
from os import path, getenv, environ
from sys import platform

from .radio import PyRadio
from .config import PyRadioConfig

PATTERN = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'

IMPLEMENTED_PLAYERS =('mpv', 'mplayer', 'cvlc')

def __configureLogger():
    logger = logging.getLogger("pyradio")
    logger.setLevel(logging.DEBUG)

    # Handler
    fh = logging.FileHandler(path.join(path.expanduser("~"), "pyradio.log"))
    fh.setLevel(logging.DEBUG)

    # create formatter
    formatter = logging.Formatter(PATTERN)

    # add formatter to ch
    fh.setFormatter(formatter)

    # add ch to logger
    logger.addHandler(fh)

def shell():
    version_too_old = False
    if sys.version_info[0] == 2:
        if sys.version_info < (2, 7):
            version_too_old = True
        elif sys.version_info.major == 3 and sys.version_info < (3, 5):
            version_too_old = True
    if version_too_old:
        print('Pyradio requires python 2.7 or 3.5+...')
        sys.exit(1)

    requested_player = ''
    parser = ArgumentParser(description="Curses based Internet radio player")
    parser.add_argument("-s", "--stations", default='',
                        help="Use specified station CSV file.")
    parser.add_argument("-p", "--play", nargs='?', default='False',
                        help="Start and play."
                        "The value is num station or empty for random.")
    parser.add_argument("-u", "--use-player", default='',
            help="Use specified player. "
            "A comma-separated list can be used to specify detection order. "
            "Supported players: mpv, mplayer, vlc.")
    parser.add_argument("-a", "--add", action='store_true',
                        help="Add station to list.")
    parser.add_argument("-ls", "--list-playlists", action='store_true',
                        help="List of available playlists in config dir.")
    parser.add_argument("-l", "--list", action='store_true',
                        help="List of available stations in a playlist.")
    parser.add_argument("-t", "--theme", default='', help="Use specified theme. ")
    parser.add_argument("-scd", "--show-config-dir", action='store_true',
                        help="Print config directory location and exit.")
    parser.add_argument("-ocd", "--open-config-dir", action='store_true',
                        help="Open config directory with default file manager.")
    parser.add_argument('--unlock', action='store_true',
                        help="Remove sessions' lock file.")
    parser.add_argument("-d", "--debug", action='store_true',
                        help="Start pyradio in debug mode.")
    args = parser.parse_args()

    sys.stdout.flush()
    pyradio_config = PyRadioConfig()

    # set window title
    if platform.startswith('win'):
        import ctypes
        try:
            if pyradio_config.locked:
                ctypes.windll.kernel32.SetConsoleTitleW("PyRadio: The Internet Radio player (Session Locked)")
            else:
                ctypes.windll.kernel32.SetConsoleTitleW("PyRadio: The Internet Radio player")
        except:
            pass
    else:
        try:
            if pyradio_config.locked:
                sys.stdout.write("\x1b]2;PyRadio: The Internet Radio player (Session Locked)\x07")
            else:
                sys.stdout.write("\x1b]2;PyRadio: The Internet Radio player\x07")
        except:
            pass

    if args.show_config_dir:
        print('PyRadio config dir: "{}"'.format(pyradio_config.stations_dir))
        sys.exit()

    if args.open_config_dir:
        open_conf_dir(pyradio_config)
        sys.exit()

    if args.list_playlists:
        pyradio_config.list_playlists()
        sys.exit()

    if args.unlock:
        lock_file = path.join(pyradio_config.stations_dir, '.lock')
        if path.exists(lock_file):
            from os import remove
            try:
                remove(lock_file)
                print('Lock file removed...')
            except:
                print('Failed to remove Lock file...')
        else:
            print('Lock file not found...')
        sys.exit()

    if args.list is False and args.add is False:
        print('Reading config...')
        ret = pyradio_config.read_config()
        if ret == -1:
            print('Error opening config: "{}"'.format(pyradio_config.config_file))
            sys.exit(1)
        elif ret == -2:
            print('Config file is malformed: "{}"'.format(pyradio_config.config_file))
            sys.exit(1)

    if args.use_player != '':
        if not platform.startswith('win'):
            requested_player = args.use_player

    if args.list is False and args.add is False:
        print('Reading playlist...')
    sys.stdout.flush()
    ret = pyradio_config.read_playlist_file(stationFile=args.stations)
    if ret < 0:
        print_playlist_selection_error(args.stations, pyradio_config, ret)

    # No need to parse the file if we add station
    # Actually we do need to do so now, so that we
    # handle 2-column vs. 3-column playlists
    if args.add:
        if sys.version_info < (3, 0):
            params = raw_input("Enter the name: "), raw_input("Enter the url: "), raw_input("Enter the encoding (leave empty for 'utf-8'): ")
        else:
            params = input("Enter the name: "), input("Enter the url: "), input("Enter the encoding (leave empty for 'utf-8'): ")
        msg = ('name', 'url')
        for i, a_param in enumerate(params):
            if i < 2:
                if a_param.strip() == '':
                    print('** Error: No {} entered. Aborting...'.format(msg[i]))
                    sys.exit(1)
        ret = pyradio_config.append_station(params, args.stations)
        if ret < 0:
            print_playlist_selection_error(args.stations, pyradio_config, ret)
        sys.exit()

    if args.list:
        header_format_string, format_string = get_format_string(pyradio_config.stations)
        header_string = header_format_string.format('[Name]','[URL]','[Encoding]')
        print(header_string)
        print(len(header_string) * '-')
        for num, a_station in enumerate(pyradio_config.stations):
            if a_station[2]:
                encoding = a_station[2]
            else:
                encoding = pyradio_config.default_encoding
            print(format_string.format(str(num+1), a_station[0], a_station[1], encoding))
        sys.exit()

    if args.debug:
        __configureLogger()
        print('Debug mode activated; printing messages to file: "~/pyradio.log"')
    else:
        ''' Refer to https://docs.python.org/3.7/howto/logging.html
            section "What happens if no configuration is provided"
        '''
        logging.raiseExceptions = False
        logging.lastResort = None

    if requested_player == '':
        requested_player = pyradio_config.player
    #else:
    #    pyradio_config.requested_player = requested_player

    if args.play == 'False':
        if args.stations == '':
            args.play = pyradio_config.default_station
    if args.play == '-1':
        args.play = 'False'

    theme_to_use = args.theme
    if not theme_to_use:
        theme_to_use = pyradio_config.theme

    # Starts the radio gui.
    pyradio = PyRadio(pyradio_config, 
            play=args.play, 
            req_player=requested_player,
            theme=theme_to_use)
    """ Setting ESCAPE key delay to 25ms
    Refer to: https://stackoverflow.com/questions/27372068/why-does-the-escape-key-have-a-delay-in-python-curses"""
    environ.setdefault('ESCDELAY', '25')
    curses.wrapper(pyradio.setup)
    if pyradio.setup_return_status:
        print('\nThank you for using PyRadio. Cheers!')
    else:
        print('\nThis terminal can not display colors.\nPyRadio cannot function in such a terminal.\n')

def print_playlist_selection_error(a_selection, cnf, ret, exit_if_malformed=True):
    if exit_if_malformed:
        if ret == -1:
            print('Error: playlist is malformed: "{}"'.format(a_selection))
            #print('Error: playlist is malformed: "{}"'.format(args.stations))
            sys.exit(1)

    if ret == -2:
        print('Error: Specified playlist not found')
        sys.exit(1)
    elif ret == -3:
        print('Error: Negative playlist number specified')
        sys.exit(1)
    elif ret == -4:
        print('Error: Specified numbered playlist not found')
        cnf.list_playlists()
        sys.exit(1)
    elif ret == -5:
        print('Error: Failed to write playlist')
        sys.exit(1)
    elif ret == -6:
        print('Error: Failed to rename playlist')
        sys.exit(1)
    elif ret == -7:
        print('Error: Playlist recovery failed!\n')
        if cnf.playlist_recovery_result == 1:
            msg = """Both a playlist file (CSV) and a playlist backup file (TXT)
exist for the selected playlist. In this case, PyRadio would
try to delete the CSV file, and then rename the TXT file to CSV.\n
Unfortunately, deleting the CSV file has failed, so you have to
manually address the issue."""
        else:
            msg = """A playlist backup file (TXT) has been found for the selected 
playlist. In this case, PyRadio would try to rename this file
to CSV.\n
Unfortunately, renaming this file has failed, so you have to
manually address the issue."""
        print(msg)
        #open_conf_dir(cnf)
        sys.exit(1)
    elif ret == -8:
        print('File type not supported')
        sys.exit(1)

def open_conf_dir(cnf):
    import subprocess
    import os
    import platform
    if platform.system().lower() == "windows":
        os.startfile(cnf.stations_dir)
    elif platform.system().lower() == "darwin":
        subprocess.Popen(["open", cnf.stations_dir])
    else:
        subprocess.Popen(["xdg-open", cnf.stations_dir])

def get_format_string(stations):
    len0 = len1 = 0
    for n in stations:
        if len(n[0]) > len0:
            len0 = len(n[0])
        if len(n[1]) > len1:
            len1 = len(n[1])
    num = len(str(len(stations)))
    format_string = '{0:>' + str(num) + '.' + str(num) + 's}. ' + '{1:' + str(len0) + '.' + str(len0) + 's} | {2:' + str(len1) + '.' + str(len1) + 's} | {3}'
    header_format_string = '{0:' + str(len0+num+2) + '.' + str(len0+num+2) + 's} | {1:' + str(len1) + '.' + str(len1) + 's} | {2}'
    return header_format_string, format_string

if __name__ == '__main__':
    shell()
